diff --git a/core/linux-raspberrypi/PKGBUILD b/core/linux-raspberrypi/PKGBUILD
index 72f6c1a..645ff2c 100644
--- a/core/linux-raspberrypi/PKGBUILD
+++ b/core/linux-raspberrypi/PKGBUILD
@@ -27,6 +27,9 @@ source=("https://github.com/raspberrypi/linux/archive/${_commit}.tar.gz"
         "ftp://teambelgium.net/bfq/patches/${pkgver:0:4}.0-${bfqver}/0003-block-bfq-add-Early-Queue-Merge-EQM-to-BFQ-${bfqver}-for-${pkgver:0:4}.0.patch"
         'https://github.com/raspberrypi/tools/raw/master/mkimage/mkknlimg'
         'config.txt'
+        'rpi900-pcf8523-i2c-register.patch'
+        'rpi900-high-speed-serial.patch'
+        'dnt900-0.3.5.tar.gz::https://github.com/mholling/dnt900/archive/v0.3.5.tar.gz'
         'cmdline.txt'
         'config.v6'
         'config.v7')
@@ -37,6 +40,9 @@ md5sums=('dd4f412940e328603fcb0aa72ac37a8e'
          '8f2743651280f5a022e541f4e95e5546'
          '405015cdeb544575c25018b3487a2e76'
          '74f15f615923836da8c5b7e9af26714d'
+         'f9719bb692f28c6e47bf1559783115ee'
+         '031562687b42f794bb62d3a4541e18d9'
+         'ec1ca90b907f05f0f614193d1a2a4da2'
          '60bc3624123c183305677097bcd56212'
          '05a8919cc2ce7211db9235b7cd82d93a'
          '3bde46c51108d218ec60fd8f5db2aa40')
@@ -63,6 +69,14 @@ prepare() {
 msg "Prepare to build"
   [[ $CARCH == "armv6h" ]] && cat "${srcdir}/config.v6" > ./.config
   [[ $CARCH == "armv7h" ]] && cat "${srcdir}/config.v7" > ./.config
+  sed -i '/CONFIG_I2C_CHARDEV=/s/m/y/' .config
+  sed -i '/CONFIG_I2C_BCM2708=/s/m/y/' .config
+  sed -i '/CONFIG_RTC_DRV_PCF8523=/s/m/y/' .config
+  sed -i 's/# CONFIG_RTC_HCTOSYS is not set/CONFIG_RTC_HCTOSYS=y/' .config
+  patch -Np1 -i "${srcdir}/rpi900-pcf8523-i2c-register.patch"
+  patch -Np1 -i "${srcdir}/rpi900-high-speed-serial.patch"
+  cp "${srcdir}/dnt900-0.3.5/dnt900.c" drivers/tty/
+  echo 'obj-y += dnt900.o' >> drivers/tty/Makefile
 
   # add pkgrel to extraversion
   sed -ri "s|^(EXTRAVERSION =)(.*)|\1 \2-${pkgrel}|" Makefile
diff --git a/core/linux-raspberrypi/rpi900-high-speed-serial.patch b/core/linux-raspberrypi/rpi900-high-speed-serial.patch
new file mode 100644
index 0000000..4836025
--- /dev/null
+++ b/core/linux-raspberrypi/rpi900-high-speed-serial.patch
@@ -0,0 +1,13 @@
+diff --git a/arch/arm/mach-bcm2708/include/mach/platform.h b/arch/arm/mach-bcm2708/include/mach/platform.h
+index 2e7e1bb..62a2ba8 100644
+--- a/arch/arm/mach-bcm2708/include/mach/platform.h
++++ b/arch/arm/mach-bcm2708/include/mach/platform.h
+@@ -221,7 +221,7 @@
+ #define PM_RSTS_HADDRF_SET                                 0x00000002
+ #define PM_RSTS_HADDRQ_SET                                 0x00000001
+ 
+-#define UART0_CLOCK      3000000
++#define UART0_CLOCK      12000000
+ 
+ #endif
+ 
diff --git a/core/linux-raspberrypi/rpi900-pcf8523-i2c-register.patch b/core/linux-raspberrypi/rpi900-pcf8523-i2c-register.patch
new file mode 100644
index 0000000..3817af5
--- /dev/null
+++ b/core/linux-raspberrypi/rpi900-pcf8523-i2c-register.patch
@@ -0,0 +1,30 @@
+diff --git a/arch/arm/mach-bcm2708/bcm2708.c b/arch/arm/mach-bcm2708/bcm2708.c
+index 13b91de..390a3ff 100644
+--- a/arch/arm/mach-bcm2708/bcm2708.c
++++ b/arch/arm/mach-bcm2708/bcm2708.c
+@@ -617,6 +617,12 @@ static struct platform_device bcm2835_thermal_device = {
+ 	.name = "bcm2835_thermal",
+ };
+ 
++static struct i2c_board_info offboard_i2c_devices[] = {
++	{
++		I2C_BOARD_INFO("pcf8523", 0x68),
++	},
++};
++
+ int __init bcm_register_device(struct platform_device *pdev)
+ {
+ 	int ret;
+@@ -754,6 +760,12 @@ void __init bcm2708_init(void)
+ 	spi_register_board_info(bcm2708_spi_devices,
+ 			ARRAY_SIZE(bcm2708_spi_devices));
+ #endif
++
++	/* test system revision to deterimine I2C bus number for GPIO header */
++	i = system_rev < 4 ? 0 : 1;
++	i2c_register_board_info(i, offboard_i2c_devices,
++				ARRAY_SIZE(offboard_i2c_devices));
++
+ }
+ 
+ static void timer_set_mode(enum clock_event_mode mode,
